{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "afc0491c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: Bio in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (1.5.8)\n",
      "Requirement already satisfied: mygene in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (3.2.2)\n",
      "Requirement already satisfied: gprofiler-official in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (1.0.0)\n",
      "Requirement already satisfied: tqdm in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (4.64.1)\n",
      "Requirement already satisfied: biopython>=1.80 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (1.81)\n",
      "Requirement already satisfied: pooch in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (1.7.0)\n",
      "Requirement already satisfied: pandas in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (1.4.4)\n",
      "Requirement already satisfied: requests in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (2.28.1)\n",
      "Requirement already satisfied: numpy in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from biopython>=1.80->Bio) (1.24.2)\n",
      "Requirement already satisfied: biothings-client>=0.2.6 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from mygene->Bio) (0.3.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pandas->Bio) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pandas->Bio) (2022.1)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pooch->Bio) (21.3)\n",
      "Requirement already satisfied: platformdirs>=2.5.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pooch->Bio) (2.5.2)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests->Bio) (1.26.11)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests->Bio) (3.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests->Bio) (2022.9.24)\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests->Bio) (2.0.4)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from packaging>=20.0->pooch->Bio) (3.0.9)\n",
      "Requirement already satisfied: six>=1.5 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from python-dateutil>=2.8.1->pandas->Bio) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install Bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "74e368bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting modlamp\n",
      "  Downloading modlamp-4.3.0.tar.gz (8.0 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.0/8.0 MB\u001b[0m \u001b[31m2.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: setuptools>=20.2.2 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (63.4.1)\n",
      "Requirement already satisfied: nose>=1.3.7 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.3.7)\n",
      "Requirement already satisfied: numpy>=1.10.4 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.21.5)\n",
      "Requirement already satisfied: scipy>=0.17.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.9.1)\n",
      "Requirement already satisfied: matplotlib>=1.5.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (3.5.2)\n",
      "Requirement already satisfied: scikit-learn>=0.18.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.0.2)\n",
      "Requirement already satisfied: pandas>=0.18.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.4.4)\n",
      "Requirement already satisfied: requests>=2.11.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (2.28.1)\n",
      "Requirement already satisfied: lxml>=3.6.4 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (4.9.1)\n",
      "Requirement already satisfied: joblib>=0.15.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.1.0)\n",
      "Collecting mysql-connector-python==8.0.17\n",
      "  Downloading mysql_connector_python-8.0.17-py2.py3-none-any.whl (345 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m345.2/345.2 kB\u001b[0m \u001b[31m26.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting protobuf>=3.0.0\n",
      "  Downloading protobuf-4.22.1-cp37-abi3-macosx_10_9_universal2.whl (397 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m397.2/397.2 kB\u001b[0m \u001b[31m4.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: python-dateutil>=2.7 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (2.8.2)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (21.3)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (9.2.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (4.25.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (1.4.2)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (3.0.9)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (0.11.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pandas>=0.18.1->modlamp) (2022.1)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (1.26.11)\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (2.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (2022.9.24)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (3.3)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from scikit-learn>=0.18.0->modlamp) (2.2.0)\n",
      "Requirement already satisfied: six>=1.5 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from python-dateutil>=2.7->matplotlib>=1.5.1->modlamp) (1.16.0)\n",
      "Building wheels for collected packages: modlamp\n",
      "  Building wheel for modlamp (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for modlamp: filename=modlamp-4.3.0-py3-none-any.whl size=173889 sha256=6adcff4fac8fc660d9f282b12ad660acd9c24ea8d6cf8e43c64c749542d8eeeb\n",
      "  Stored in directory: /Users/ggcl7/Library/Caches/pip/wheels/fc/09/85/aefb43b4f79b6323733e8276c157c4851409b65f77d7248ace\n",
      "Successfully built modlamp\n",
      "Installing collected packages: protobuf, mysql-connector-python, modlamp\n",
      "Successfully installed modlamp-4.3.0 mysql-connector-python-8.0.17 protobuf-4.22.1\n"
     ]
    }
   ],
   "source": [
    "!pip install modlamp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c40e3ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: tqdm in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (4.64.1)\r\n"
     ]
    }
   ],
   "source": [
    "!pip install tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5e073c2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis as PA\n",
    "from modlamp.descriptors import PeptideDescriptor, GlobalDescriptor\n",
    "from sklearn.model_selection import train_test_split\n",
    "import pandas as pd\n",
    "import os, re, math, platform\n",
    "from tqdm import tqdm\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "from Bio import Seq, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "76593489",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:469: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:471: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:473: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:469: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:471: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:473: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "/var/folders/19/51xjwpd57m1g5zzk1p6ps9g40000gn/T/ipykernel_6723/3287707131.py:469: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  if n is 1:\n",
      "/var/folders/19/51xjwpd57m1g5zzk1p6ps9g40000gn/T/ipykernel_6723/3287707131.py:471: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  elif n is 2:\n",
      "/var/folders/19/51xjwpd57m1g5zzk1p6ps9g40000gn/T/ipykernel_6723/3287707131.py:473: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  elif n is 3:\n"
     ]
    }
   ],
   "source": [
    "# Miscellaneous\n",
    "_AALetter = ['A', 'C', 'D', 'E', 'F', 'G', 'H',\n",
    "             'I', 'K', 'L', 'M', 'N', 'P', 'Q',\n",
    "             'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "n_gram statistics\n",
    "\"\"\"\n",
    "\n",
    "_AALetter = ['A', 'C', 'D', 'E', 'F', 'G', 'H', \n",
    "            'I', 'K', 'L', 'M', 'N', 'P', 'Q', \n",
    "            'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "\n",
    "def get_aan_corpus(n=2):\n",
    "    '''\n",
    "    Get AA corpus of n_gram (e.g. Di, Tri, etc.)\n",
    "    Output: AA n_gram corpus ((e.g. Di:400, Tri:3000, etc.))\n",
    "    '''\n",
    "    n_corpus = []\n",
    "    if n <= 2:\n",
    "        for i in _AALetter:\n",
    "            for j in _AALetter:\n",
    "               n_corpus.append(\"{}{}\".format(i, j))\n",
    "        return n_corpus\n",
    "    for i in get_aan_corpus(n - 1):\n",
    "        for j in _AALetter:\n",
    "            n_corpus.append(\"{}{}\".format(i, j))\n",
    "    return n_corpus\n",
    "\n",
    "\n",
    "def get_ngram_counts(seq, n=2):\n",
    "    '''\n",
    "    Get n_gram statistics\n",
    "    Input: peptide sequence and n\n",
    "    Ouput: n_gram statistic (dictionary) {A.A Corp: Counts}\n",
    "    '''\n",
    "    # Get the name of ngram feature\n",
    "    if n == 2:\n",
    "        prefix = 'DPC'\n",
    "    elif n == 3:\n",
    "        prefix = 'TPC'\n",
    "    else:\n",
    "        prefix = '{}gram'.format(n)\n",
    "\n",
    "    ngrams = [seq[i: i + n] for i in range(len(seq) - n + 1)]\n",
    "    n_corpus = get_aan_corpus(n)\n",
    "    ngram_stat = {}\n",
    "    for aa_ng in n_corpus:\n",
    "        ngram_stat['{}|{}'.format(prefix, aa_ng)] = ngrams.count(aa_ng) / len(ngrams) * 100\n",
    "    return ngram_stat\n",
    "\n",
    "\n",
    "def minSequenceLength(fastas):\n",
    "    minLen = 10000\n",
    "    for i in fastas:\n",
    "        if minLen > len(i[1]):\n",
    "            minLen = len(i[1])\n",
    "    return minLen\n",
    "\n",
    "\n",
    "def minSequenceLengthWithNormalAA(fastas):\n",
    "    minLen = 10000\n",
    "    for i in fastas:\n",
    "        if minLen > len(re.sub('-', '', i[1])):\n",
    "            minLen = len(re.sub('-', '', i[1]))\n",
    "    return minLen\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "    input.fasta:      the input protein sequence file in fasta format.\n",
    "    k_space:          the gap of two amino acids, integer, defaule: 5\n",
    "    output:           the encoding file, default: 'encodings.tsv'\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def generateGroupPairs(groupKey):\n",
    "    gPair = {}\n",
    "    for key1 in groupKey:\n",
    "        for key2 in groupKey:\n",
    "            gPair['CKSAAGP|'+key1+'.'+key2] = 0\n",
    "    return gPair\n",
    "\n",
    "\n",
    "def cksaagp(fastas, gap = 5, **kw):\n",
    "    if gap < 0:\n",
    "        print('Error: the gap should be equal or greater than zero' + '\\n\\n')\n",
    "        return 0\n",
    "\n",
    "    if minSequenceLength(fastas) < gap+2:\n",
    "        print('Error: all the sequence length should be greater than the (gap value) + 2 = ' + str(gap+2) + '\\n\\n')\n",
    "        return 0\n",
    "\n",
    "    group = {'aliphatic': 'GAVLMI',\n",
    "             'aromatic': 'FYW',\n",
    "             'postivecharge': 'KRH',\n",
    "             'negativecharge': 'DE',\n",
    "             'uncharge': 'STCPNQ'}\n",
    "\n",
    "    AA = 'ARNDCQEGHILKMFPSTWYV'\n",
    "\n",
    "    groupKey = group.keys()\n",
    "\n",
    "    index = {}\n",
    "    for key in groupKey:\n",
    "        for aa in group[key]:\n",
    "            index[aa] = key\n",
    "\n",
    "    gPairIndex = []\n",
    "    for key1 in groupKey:\n",
    "        for key2 in groupKey:\n",
    "            gPairIndex.append('CKSAAGP|'+key1+'.'+key2)\n",
    "\n",
    "    encodings = []\n",
    "    header = ['#']\n",
    "    for g in range(gap + 1):\n",
    "        for p in gPairIndex:\n",
    "            header.append(p+'.gap'+str(g))\n",
    "    encodings.append(header)\n",
    "\n",
    "    for i in fastas:\n",
    "        name, sequence = i[0], re.sub('-', '', i[1])\n",
    "        code = [name]\n",
    "        for g in range(gap + 1):\n",
    "            gPair = generateGroupPairs(groupKey)\n",
    "            sum = 0\n",
    "            for p1 in range(len(sequence)):\n",
    "                p2 = p1 + g + 1\n",
    "                if p2 < len(sequence) and sequence[p1] in AA and sequence[p2] in AA:\n",
    "                    gPair['CKSAAGP|'+index[sequence[p1]]+'.'+index[sequence[p2]]] = gPair['CKSAAGP|'+index[sequence[p1]]+'.'+index[sequence[p2]]] + 1\n",
    "                    sum = sum + 1\n",
    "\n",
    "            if sum == 0:\n",
    "                for gp in gPairIndex:\n",
    "                    code.append(0)\n",
    "            else:\n",
    "                for gp in gPairIndex:\n",
    "                    code.append(gPair[gp] / sum)\n",
    "\n",
    "        encodings.append(code)\n",
    "\n",
    "    return encodings\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "    input.fasta:      the input protein sequence file in fasta format.\n",
    "    lambda:           the lambda value, integer, defaule: 30\n",
    "    output:           the encoding file, default: 'encodings.tsv'\n",
    "\"\"\"\n",
    "\n",
    "def Rvalue(aa1, aa2, AADict, Matrix):\n",
    "    return sum([(Matrix[i][AADict[aa1]] - Matrix[i][AADict[aa2]]) ** 2 for i in range(len(Matrix))]) / len(Matrix)\n",
    "\n",
    "\n",
    "def paac(fastas, lambdaValue=30, w=0.05, **kw):\n",
    "    if minSequenceLengthWithNormalAA(fastas) < lambdaValue + 1:\n",
    "        print('Error: all the sequence length should be larger than the lambdaValue+1: ' + str(lambdaValue + 1) + '\\n\\n')\n",
    "        return 0\n",
    "\n",
    "    dataFile = re.sub('codes$', '', os.path.split(os.path.realpath('__file__'))[0]) + r'\\data\\PAAC.txt' if platform.system() == 'Windows' else re.sub('codes$', '', os.path.split(os.path.realpath('__file__'))[0]) + '/data/PAAC.txt'\n",
    "    with open(dataFile) as f:\n",
    "        records = f.readlines()\n",
    "    AA = ''.join(records[0].rstrip().split()[1:])\n",
    "    AADict = {}\n",
    "    for i in range(len(AA)):\n",
    "        AADict[AA[i]] = i\n",
    "    AAProperty = []\n",
    "    AAPropertyNames = []\n",
    "    for i in range(1, len(records)):\n",
    "        array = records[i].rstrip().split() if records[i].rstrip() != '' else None\n",
    "        AAProperty.append([float(j) for j in array[1:]])\n",
    "        AAPropertyNames.append(array[0])\n",
    "\n",
    "    AAProperty1 = []\n",
    "    for i in AAProperty:\n",
    "        meanI = sum(i) / 20\n",
    "        fenmu = math.sqrt(sum([(j-meanI)**2 for j in i])/20)\n",
    "        AAProperty1.append([(j-meanI)/fenmu for j in i])\n",
    "\n",
    "    encodings = []\n",
    "    header = ['#']\n",
    "    for aa in AA:\n",
    "        header.append('PAAC|' + aa)\n",
    "    for n in range(1, lambdaValue + 1):\n",
    "        header.append('PAAC|lambda' + str(n))\n",
    "    encodings.append(header)\n",
    "\n",
    "    for i in fastas:\n",
    "        name, sequence = i[0], re.sub('-', '', i[1])\n",
    "        code = [name]\n",
    "        theta = []\n",
    "        for n in range(1, lambdaValue + 1):\n",
    "            theta.append(\n",
    "                sum([Rvalue(sequence[j], sequence[j + n], AADict, AAProperty1) for j in range(len(sequence) - n)]) / (\n",
    "                len(sequence) - n))\n",
    "        myDict = {}\n",
    "        for aa in AA:\n",
    "            myDict[aa] = sequence.count(aa)\n",
    "        code = code + [myDict[aa] / (1 + w * sum(theta)) for aa in AA]\n",
    "        code = code + [(w * j) / (1 + w * sum(theta)) for j in theta]\n",
    "        encodings.append(code)\n",
    "    return encodings\n",
    "\n",
    "\n",
    "def GAAC(fastas, **kw):\n",
    "\tgroup = {\n",
    "\t\t'alphatic': 'GAVLMI',\n",
    "\t\t'aromatic': 'FYW',\n",
    "\t\t'postivecharge': 'KRH',\n",
    "\t\t'negativecharge': 'DE',\n",
    "\t\t'uncharge': 'STCPNQ'\n",
    "\t}\n",
    "\n",
    "\tgroupKey = group.keys()\n",
    "\n",
    "\tencodings = []\n",
    "\theader = ['#']\n",
    "\tfor key in groupKey:\n",
    "\t\theader.append(\"GAAC|\"+key)\n",
    "\tencodings.append(header)\n",
    "\n",
    "\tfor i in fastas:\n",
    "\t\tname, sequence = i[0], re.sub('-', '', i[1])\n",
    "\t\tcode = [name]\n",
    "\t\tcount = Counter(sequence)\n",
    "\t\tmyDict = {}\n",
    "\t\tfor key in groupKey:\n",
    "\t\t\tfor aa in group[key]:\n",
    "\t\t\t\tmyDict[key] = myDict.get(key, 0) + count[aa]\n",
    "\n",
    "\t\tfor key in groupKey:\n",
    "\t\t\tcode.append(myDict[key]/len(sequence))\n",
    "\t\tencodings.append(code)\n",
    "\n",
    "\treturn encodings\n",
    "\n",
    "\n",
    "def GDPC(fastas, **kw):\n",
    "\tgroup = {\n",
    "\t\t'alphaticr': 'GAVLMI',\n",
    "\t\t'aromatic': 'FYW',\n",
    "\t\t'postivecharger': 'KRH',\n",
    "\t\t'negativecharger': 'DE',\n",
    "\t\t'uncharger': 'STCPNQ'\n",
    "\t}\n",
    "\n",
    "\tgroupKey = group.keys()\n",
    "\tbaseNum = len(groupKey)\n",
    "\tdipeptide = [g1+'.'+g2 for g1 in groupKey for g2 in groupKey]\n",
    "\n",
    "\tindex = {}\n",
    "\tfor key in groupKey:\n",
    "\t\tfor aa in group[key]:\n",
    "\t\t\tindex[aa] = key\n",
    "\n",
    "\tencodings = []\n",
    "\theader = ['#'] + ['GDPC|'+dipname for dipname in dipeptide]\n",
    "\tencodings.append(header)\n",
    "\n",
    "\tfor i in fastas:\n",
    "\t\tname, sequence = i[0], re.sub('-', '', i[1])\n",
    "\n",
    "\t\tcode = [name]\n",
    "\t\tmyDict = {}\n",
    "\t\tfor t in dipeptide:\n",
    "\t\t\tmyDict[t] = 0\n",
    "\n",
    "\t\tsum = 0\n",
    "\t\tfor j in range(len(sequence) - 2 + 1):\n",
    "\t\t\tmyDict[index[sequence[j]]+'.'+index[sequence[j+1]]] = myDict[index[sequence[j]]+'.'+index[sequence[j+1]]] + 1\n",
    "\t\t\tsum = sum +1\n",
    "\n",
    "\t\tif sum == 0:\n",
    "\t\t\tfor t in dipeptide:\n",
    "\t\t\t\tcode.append(0)\n",
    "\t\telse:\n",
    "\t\t\tfor t in dipeptide:\n",
    "\t\t\t\tcode.append(myDict[t]/sum)\n",
    "\t\tencodings.append(code)\n",
    "\n",
    "\treturn encodings\n",
    "\n",
    "\n",
    "def GTPC(fastas, **kw):\n",
    "\tgroup = {\n",
    "\t\t'alphaticr': 'GAVLMI',\n",
    "\t\t'aromatic': 'FYW',\n",
    "\t\t'postivecharger': 'KRH',\n",
    "\t\t'negativecharger': 'DE',\n",
    "\t\t'uncharger': 'STCPNQ'\n",
    "\t}\n",
    "\n",
    "\tgroupKey = group.keys()\n",
    "\tbaseNum = len(groupKey)\n",
    "\ttriple = [g1+'.'+g2+'.'+g3 for g1 in groupKey for g2 in groupKey for g3 in groupKey]\n",
    "\n",
    "\tindex = {}\n",
    "\tfor key in groupKey:\n",
    "\t\tfor aa in group[key]:\n",
    "\t\t\tindex[aa] = key\n",
    "\n",
    "\tencodings = []\n",
    "\theader = ['#'] + ['GTPC|'+tname for tname in triple]\n",
    "\tencodings.append(header)\n",
    "\n",
    "\tfor i in fastas:\n",
    "\t\tname, sequence = i[0], re.sub('-', '', i[1])\n",
    "\n",
    "\t\tcode = [name]\n",
    "\t\tmyDict = {}\n",
    "\t\tfor t in triple:\n",
    "\t\t\tmyDict[t] = 0\n",
    "\n",
    "\t\tsum = 0\n",
    "\t\tfor j in range(len(sequence) - 3 + 1):\n",
    "\t\t\tmyDict[index[sequence[j]]+'.'+index[sequence[j+1]]+'.'+index[sequence[j+2]]] = myDict[index[sequence[j]]+'.'+index[sequence[j+1]]+'.'+index[sequence[j+2]]] + 1\n",
    "\t\t\tsum = sum +1\n",
    "\n",
    "\t\tif sum == 0:\n",
    "\t\t\tfor t in triple:\n",
    "\t\t\t\tcode.append(0)\n",
    "\t\telse:\n",
    "\t\t\tfor t in triple:\n",
    "\t\t\t\tcode.append(myDict[t]/sum)\n",
    "\t\tencodings.append(code)\n",
    "\n",
    "\treturn encodings\n",
    "\n",
    "'''\n",
    "Insert Iso_electric Point and net_charge(neutral) feature to the sequence data_frame\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., iep, net_charge}\n",
    "'''\n",
    "\n",
    "\n",
    "def insert_phycs(seq_df):\n",
    "    seq_df = seq_df.copy()\n",
    "    #  Function for compute Isoelectric Point or net_charge of peptide\n",
    "    def get_ieq_nc(seq, is_iep=True):\n",
    "        protparam = PA(seq)\n",
    "        return protparam.isoelectric_point() if is_iep else protparam.charge_at_pH(7.0)\n",
    "\n",
    "    # Calculating IsoElectricPoints and NeutralCharge\n",
    "    data_size = seq_df.size\n",
    "    seq_df['PHYC|IEP'] = list(map(get_ieq_nc, seq_df['Sequence'], [True] * data_size))  # IsoElectricPoints\n",
    "    seq_df['PHYC|Net Charge'] = list(map(get_ieq_nc, seq_df['Sequence'], [False] * data_size))  # Charge(Neutral)\n",
    "\n",
    "    # Calculating hydrophobic moment (My assume all peptides are alpha-helix)\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'eisenberg')\n",
    "    descrpt.calculate_moment(window=1000, angle=100, modality='max')\n",
    "    seq_df['PHYC|Hydrophobic Moment'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating \"Hopp-Woods\" hydrophobicity\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'hopp-woods')\n",
    "    descrpt.calculate_global()\n",
    "    seq_df['PHYC|Hydrophobicity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Energy of Transmembrane Propensity\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'tm_tend')\n",
    "    descrpt.calculate_global()\n",
    "    seq_df['PHYC|Transmembrane Propensity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Aromaticity\n",
    "    descrpt = GlobalDescriptor(seq_df['Sequence'].values)\n",
    "    descrpt.aromaticity()\n",
    "    seq_df['PHYC|Aromacity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Levitt_alpha_helical Propensity\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'levitt_alpha')\n",
    "    descrpt.calculate_global()\n",
    "    seq_df['PHYC|Alpha Helical Propensity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Aliphatic Index\n",
    "    descrpt = GlobalDescriptor(seq_df['Sequence'].values)\n",
    "    descrpt.aliphatic_index()\n",
    "    seq_df['PHYC|Aliphatic Index'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Boman Index\n",
    "    descrpt = GlobalDescriptor(seq_df['Sequence'].values)\n",
    "    descrpt.boman_index()\n",
    "    seq_df['PHYC|Boman Index'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "'''\n",
    "Insert Amino acid composition to the sequence data_frame\n",
    "Input: sequence data_frame {IDx: Seq_x}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., AAC_Ax ... AAC_Yx}\n",
    "'''\n",
    "\n",
    "\n",
    "def insert_aac(seq_df):\n",
    "    seq_df = seq_df.copy()\n",
    "    # Compute AAC for peptide in specific A.A\n",
    "    def get_aac(seq, aa):\n",
    "        return seq.count(aa) / len(seq) * 100\n",
    "\n",
    "    # processing data_frame\n",
    "    data_size = seq_df.size\n",
    "    for ll in _AALetter:\n",
    "        seq_df['AAC|{}'.format(ll)] = list(map(get_aac, seq_df['Sequence'], [ll] * data_size))\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "'''\n",
    "Insert n_grams Descriptor to the sequence data_frame\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., ngram_(1), .., ngram(20 ** n)}\n",
    "'''\n",
    "\n",
    "\n",
    "def insert_ngrams(seq_df, n=2):\n",
    "    seq_df = seq_df.copy()\n",
    "    data_size = seq_df.size\n",
    "\n",
    "    ngrams_df = list(map(get_ngram_counts, seq_df['Sequence'], [n] * data_size))\n",
    "    ngrams_df = pd.DataFrame(ngrams_df)  # Convert ngrams features to pd.DataFrame\n",
    "    seq_df = pd.concat([seq_df, ngrams_df], axis=1)\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Insert CKSAAGP Descriptor to the sequence data_frame\n",
    "(Composition of k-spaced amino acid pairs)\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., CKSAAGP(0), ..., CKSAAGP(m)}\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def insert_cksaagp(seq_df, gap=2):\n",
    "    seq_df.copy()\n",
    "    fastas = [[idx, seq] for idx, seq in zip(seq_df['Id'], seq_df['Sequence'])]\n",
    "    encoding = cksaagp(fastas, gap=gap)\n",
    "    encoding = pd.DataFrame(encoding[1:], columns=encoding[0])\n",
    "    seq_df = pd.concat([seq_df, encoding.iloc[:, 1:]], axis=1)\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Insert PAAC Descriptor to the sequence data_frame\n",
    "(Pseudo Amino Acid Composition)\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., CKSAAGP(0), ..., CKSAAGP(m)}\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def insert_paac(seq_df, lamb=3, w=0.1):\n",
    "    seq_df = seq_df.copy()\n",
    "    fastas = [[idx, seq] for idx, seq in zip(seq_df['Id'], seq_df['Sequence'])]\n",
    "    encoding = paac(fastas, lambdaValue=lamb, w=w)\n",
    "    encoding = pd.DataFrame(encoding[1:], columns=encoding[0])\n",
    "    seq_df = pd.concat([seq_df, encoding.iloc[:, 1:]], axis=1)\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Insert Grouped n-gram Descriptor to the sequence data_frame\n",
    "(Pseudo Amino Acid Composition)\n",
    "Input: sequence data_frame {IDx: Seq_x, ...}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., CKSAAGP(0), ..., CKSAAGP(m)}\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def insert_gngram(seq_df, n=1):\n",
    "    seq_df = seq_df.copy()\n",
    "    fastas = [[idx, seq] for idx, seq in zip(seq_df['Id'], seq_df['Sequence'])]\n",
    "    # encoding = paac(fastas, lambdaValue=lamb, w=w)\n",
    "    if n is 1:\n",
    "        encoding = GAAC(fastas)\n",
    "    elif n is 2:\n",
    "        encoding = GDPC(fastas)\n",
    "    elif n is 3:\n",
    "        encoding = GTPC(fastas)\n",
    "    else:\n",
    "        raise Warning(\"Invalid n-grams, no features added\")\n",
    "    encoding = pd.DataFrame(encoding[1:], columns=encoding[0])\n",
    "    seq_df = pd.concat([seq_df, encoding.iloc[:, 1:]], axis=1)\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "def construct_features(seq_df, paaclamb=3, paacw=0.5):\n",
    "    \"\"\"\n",
    "    Construct Features for the AVPIden. We first investigated physiochemical feautres, AAC features, DiC features, \n",
    "    CKSAAGP features, PAAC features, and PHYC features.\n",
    "    Parameters are pre-set according to the sequence identities.\n",
    "    \"\"\"\n",
    "    seq_df = insert_aac(seq_df)\n",
    "#     seq_df = insert_ngrams(seq_df, n=2)\n",
    "#     seq_df = insert_gngram(seq_df, n=1)\n",
    "    seq_df = insert_cksaagp(seq_df, gap=2) # As the maximum motif length = 5.\n",
    "    seq_df = insert_paac(seq_df, lamb=paaclamb, w=paacw)\n",
    "#     seq_df = insert_phycs(seq_df)\n",
    "\n",
    "    return seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a2d488e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_fasta(df, file_path, abbr_columns=None):\n",
    "    \"\"\"\n",
    "    Save dataframe to a .fasta file, the df should contain at least columns named \"Id\" and \"Sequence\"\n",
    "    \n",
    "    df: dataframe for saving .fasta\n",
    "    file_path: path(string) for saving the fasta file\n",
    "    abbr_columns: string columns for adding abbreviations. Multiple abbr are splited by '|'.\n",
    "    \"\"\"\n",
    "    Seqrecords = [SeqIO.SeqRecord(id=row['Id'], \n",
    "                              seq=Seq.Seq(row['Sequence']), \n",
    "                              description='|'.join(row[abbr_columns] if abbr_columns is not None else \"\")) \\\n",
    "             for idn, row in df.iterrows()]\n",
    "    with open(file_path, 'w+') as fhandle:\n",
    "        SeqIO.write(Seqrecords, fhandle, \"fasta-2line\")\n",
    "        print(\"Saved {:d} sequences.\".format(len(Seqrecords)))\n",
    "\n",
    "\n",
    "def read_fasta(file_path):\n",
    "    f = open(file_path, 'r', encoding='utf-8')\n",
    "    fasta_list = np.array(f.readlines())\n",
    "    # print(fasta_list)\n",
    "    new_fasta = []\n",
    "    for flag in range(0, len(fasta_list), 2):\n",
    "        fasta_str = [fasta_list[flag].strip('\\n').strip(), fasta_list[flag + 1].strip('\\n').strip()]\n",
    "        new_fasta.append(fasta_str)\n",
    "    # print(new_fasta)\n",
    "    seq_df = pd.DataFrame(data=new_fasta, columns=['Id', 'Sequence'])\n",
    "    # print(new_fasta1.shape)\n",
    "    return seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3edcc430",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_fasta = read_fasta(\"/Users/ggcl7/Desktop/AIP_data/data/test.txt\")\n",
    "seq_feature = construct_features(seq_fasta)\n",
    "seq_feature.to_csv(\"/Users/ggcl7/Desktop/AIP_data/data/test_feature.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "24470880",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>AAC|A</th>\n",
       "      <th>AAC|C</th>\n",
       "      <th>AAC|D</th>\n",
       "      <th>AAC|E</th>\n",
       "      <th>AAC|F</th>\n",
       "      <th>AAC|G</th>\n",
       "      <th>AAC|H</th>\n",
       "      <th>AAC|I</th>\n",
       "      <th>...</th>\n",
       "      <th>PAAC|F</th>\n",
       "      <th>PAAC|P</th>\n",
       "      <th>PAAC|S</th>\n",
       "      <th>PAAC|T</th>\n",
       "      <th>PAAC|W</th>\n",
       "      <th>PAAC|Y</th>\n",
       "      <th>PAAC|V</th>\n",
       "      <th>PAAC|lambda1</th>\n",
       "      <th>PAAC|lambda2</th>\n",
       "      <th>PAAC|lambda3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;pos</td>\n",
       "      <td>MTEQQWNFAGIEAAASAIQG</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.298077</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.298077</td>\n",
       "      <td>0.298077</td>\n",
       "      <td>0.298077</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.239905</td>\n",
       "      <td>0.223799</td>\n",
       "      <td>0.238220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;pos.1</td>\n",
       "      <td>QVLAIAADYLANDAE</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.333333</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.289008</td>\n",
       "      <td>0.289008</td>\n",
       "      <td>0.221499</td>\n",
       "      <td>0.245955</td>\n",
       "      <td>0.243538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;pos.2</td>\n",
       "      <td>FFVRSIGSKGGKLAAGKYTDAVTV</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.333333</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.166667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.447980</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.447980</td>\n",
       "      <td>0.447980</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.223990</td>\n",
       "      <td>0.671971</td>\n",
       "      <td>0.235708</td>\n",
       "      <td>0.297267</td>\n",
       "      <td>0.243034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;pos.3</td>\n",
       "      <td>MQKYMQIVRYLTILITLI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>22.222222</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.592945</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.592945</td>\n",
       "      <td>0.296472</td>\n",
       "      <td>0.240278</td>\n",
       "      <td>0.310622</td>\n",
       "      <td>0.152627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;pos.4</td>\n",
       "      <td>LTMTASHTRLSCDCDDKFYD</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>10.0</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.273695</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.547390</td>\n",
       "      <td>0.821085</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.273695</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.244324</td>\n",
       "      <td>0.222621</td>\n",
       "      <td>0.259360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>&gt;neg.350</td>\n",
       "      <td>TGSAVGRGIEDSLTI</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.333333</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.481584</td>\n",
       "      <td>0.481584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.240792</td>\n",
       "      <td>0.242718</td>\n",
       "      <td>0.274679</td>\n",
       "      <td>0.241811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>&gt;neg.351</td>\n",
       "      <td>ALMPMNMAPITGRALEPYIS</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.876868</td>\n",
       "      <td>0.292289</td>\n",
       "      <td>0.292289</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292289</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.267043</td>\n",
       "      <td>0.245175</td>\n",
       "      <td>0.195492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>&gt;neg.352</td>\n",
       "      <td>VKIEYSGTNNKTMAV</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.279715</td>\n",
       "      <td>0.559431</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.279715</td>\n",
       "      <td>0.559431</td>\n",
       "      <td>0.270036</td>\n",
       "      <td>0.167191</td>\n",
       "      <td>0.283057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>708</th>\n",
       "      <td>&gt;neg.353</td>\n",
       "      <td>VNTSLVLAFSEVIHQ</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.347509</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.695019</td>\n",
       "      <td>0.347509</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.042528</td>\n",
       "      <td>0.200893</td>\n",
       "      <td>0.221202</td>\n",
       "      <td>0.230396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>&gt;neg.354</td>\n",
       "      <td>TYAPVVMKPKPQGPV</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.306941</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.326735</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.326735</td>\n",
       "      <td>0.980206</td>\n",
       "      <td>0.241478</td>\n",
       "      <td>0.156285</td>\n",
       "      <td>0.275501</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>710 rows × 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Id                  Sequence      AAC|A  AAC|C      AAC|D  \\\n",
       "0        >pos      MTEQQWNFAGIEAAASAIQG  25.000000    0.0   0.000000   \n",
       "1      >pos.1           QVLAIAADYLANDAE  33.333333    0.0  13.333333   \n",
       "2      >pos.2  FFVRSIGSKGGKLAAGKYTDAVTV  12.500000    0.0   4.166667   \n",
       "3      >pos.3        MQKYMQIVRYLTILITLI   0.000000    0.0   0.000000   \n",
       "4      >pos.4      LTMTASHTRLSCDCDDKFYD   5.000000   10.0  20.000000   \n",
       "..        ...                       ...        ...    ...        ...   \n",
       "705  >neg.350           TGSAVGRGIEDSLTI   6.666667    0.0   6.666667   \n",
       "706  >neg.351      ALMPMNMAPITGRALEPYIS  15.000000    0.0   0.000000   \n",
       "707  >neg.352           VKIEYSGTNNKTMAV   6.666667    0.0   0.000000   \n",
       "708  >neg.353           VNTSLVLAFSEVIHQ   6.666667    0.0   0.000000   \n",
       "709  >neg.354           TYAPVVMKPKPQGPV   6.666667    0.0   0.000000   \n",
       "\n",
       "         AAC|E     AAC|F      AAC|G     AAC|H      AAC|I  ...    PAAC|F  \\\n",
       "0    10.000000  5.000000  10.000000  0.000000  10.000000  ...  0.298077   \n",
       "1     6.666667  0.000000   0.000000  0.000000   6.666667  ...  0.000000   \n",
       "2     0.000000  8.333333  16.666667  0.000000   4.166667  ...  0.447980   \n",
       "3     0.000000  0.000000   0.000000  0.000000  22.222222  ...  0.000000   \n",
       "4     0.000000  5.000000   0.000000  5.000000   0.000000  ...  0.273695   \n",
       "..         ...       ...        ...       ...        ...  ...       ...   \n",
       "705   6.666667  0.000000  20.000000  0.000000  13.333333  ...  0.000000   \n",
       "706   5.000000  0.000000   5.000000  0.000000  10.000000  ...  0.000000   \n",
       "707   6.666667  0.000000   6.666667  0.000000   6.666667  ...  0.000000   \n",
       "708   6.666667  6.666667   0.000000  6.666667   6.666667  ...  0.347509   \n",
       "709   0.000000  0.000000   6.666667  0.000000   0.000000  ...  0.000000   \n",
       "\n",
       "       PAAC|P    PAAC|S    PAAC|T    PAAC|W    PAAC|Y    PAAC|V  PAAC|lambda1  \\\n",
       "0    0.000000  0.298077  0.298077  0.298077  0.000000  0.000000      0.239905   \n",
       "1    0.000000  0.000000  0.000000  0.000000  0.289008  0.289008      0.221499   \n",
       "2    0.000000  0.447980  0.447980  0.000000  0.223990  0.671971      0.235708   \n",
       "3    0.000000  0.000000  0.592945  0.000000  0.592945  0.296472      0.240278   \n",
       "4    0.000000  0.547390  0.821085  0.000000  0.273695  0.000000      0.244324   \n",
       "..        ...       ...       ...       ...       ...       ...           ...   \n",
       "705  0.000000  0.481584  0.481584  0.000000  0.000000  0.240792      0.242718   \n",
       "706  0.876868  0.292289  0.292289  0.000000  0.292289  0.000000      0.267043   \n",
       "707  0.000000  0.279715  0.559431  0.000000  0.279715  0.559431      0.270036   \n",
       "708  0.000000  0.695019  0.347509  0.000000  0.000000  1.042528      0.200893   \n",
       "709  1.306941  0.000000  0.326735  0.000000  0.326735  0.980206      0.241478   \n",
       "\n",
       "     PAAC|lambda2  PAAC|lambda3  \n",
       "0        0.223799      0.238220  \n",
       "1        0.245955      0.243538  \n",
       "2        0.297267      0.243034  \n",
       "3        0.310622      0.152627  \n",
       "4        0.222621      0.259360  \n",
       "..            ...           ...  \n",
       "705      0.274679      0.241811  \n",
       "706      0.245175      0.195492  \n",
       "707      0.167191      0.283057  \n",
       "708      0.221202      0.230396  \n",
       "709      0.156285      0.275501  \n",
       "\n",
       "[710 rows x 120 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0da4ddc9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5f28c7b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
